File: togo/internal/format/javadoc/lexer.go
// Port of src/format/javadoc/lexer.ts (which ports gjf's JavadocLexer).
//
// Lexes a classic `/** ... */` javadoc comment. The Markdown `///` path is
// deferred; classic is always assumed here.

package javadoc

import (
	"fmt"
	"regexp"
	"strings"
)

type lexError struct{ msg string }

func (e *lexError) Error() string { return e.msg }

type ctx int

const (
	ctxHTMLPre ctx = iota
	ctxHTMLCode
	ctxMarkdownCode
	ctxTable
	ctxSnippet
	ctxBrace
	ctxInlineTag
)

var (
	nonUnixLineEnding = regexp.MustCompile(`\r\n?`)
	reClassicNewline  = regexp.MustCompile(`^[ \t]*\n[ \t]*[*]?[ \t]?`)
	reFooterTag       = regexp.MustCompile(`^@(?:param\s+<\w+>|[a-z]\w*)`)
	reMoeBegin        = regexp.MustCompile(`^<!--\s*MOE:begin_intracomment_strip\s*-->`)
	reMoeEnd          = regexp.MustCompile(`^<!--\s*MOE:end_intracomment_strip\s*-->`)
	reHTMLComment     = regexp.MustCompile(`(?s)^<!--.*?-->`)
	reSnippetTagOpen  = regexp.MustCompile(`^[{]@snippet\b`)
	reInlineTagOpen   = regexp.MustCompile(`^[{]@\w*`)
	reClassicLiteral  = regexp.MustCompile(`(?s)^.[^ \t\n@<{}*]*`)

	rePreOpen         = openTagRe("pre")
	rePreClose        = closeTagRe("pre")
	reCodeOpen        = openTagRe("code")
	reCodeClose       = closeTagRe("code")
	reTableOpen       = openTagRe("table")
	reTableClose      = closeTagRe("table")
	reListOpen        = openTagRe("ul|ol|dl")
	reListClose       = closeTagRe("ul|ol|dl")
	reListItemOpen    = openTagRe("li|dt|dd")
	reListItemClose   = closeTagRe("li|dt|dd")
	reHeaderOpen      = openTagRe("h[1-6]")
	reHeaderClose     = closeTagRe("h[1-6]")
	reParagraphOpen   = openTagRe("p")
	reParagraphClose  = closeTagRe("p")
	reBlockquoteOpen  = openTagRe("blockquote")
	reBlockquoteClose = closeTagRe("blockquote")
	reBr              = openTagRe("br")
)

func openTagRe(name string) *regexp.Regexp {
	return regexp.MustCompile(fmt.Sprintf(`(?i)^<(?:%s)\b[^>]*>`, name))
}
func closeTagRe(name string) *regexp.Regexp {
	return regexp.MustCompile(fmt.Sprintf(`(?i)^</(?:%s)\b[^>]*>`, name))
}

var (
	tagContexts        = []ctx{ctxSnippet, ctxInlineTag}
	braceContexts      = []ctx{ctxSnippet, ctxInlineTag, ctxBrace}
	preserveFormatting = []ctx{ctxHTMLPre, ctxTable, ctxHTMLCode, ctxSnippet}
)

// lex lexes a `/** ... */` comment (including delimiters) into tokens.
func lex(input string) ([]Token, error) {
	input = nonUnixLineEnding.ReplaceAllString(input, "\n")
	if !strings.HasPrefix(input, "/**") || !strings.HasSuffix(input, "*/") || len(input) <= 4 {
		return nil, &lexError{"not a /** */ comment: " + input}
	}
	body := input[3 : len(input)-2]
	l := &javadocLexer{input: newCharStream(body)}
	return l.generateTokens()
}

type javadocLexer struct {
	input               *charStream
	contextStack        nestingStack[ctx]
	somethingSinceNewln bool
}

func (l *javadocLexer) generateTokens() ([]Token, error) {
	tokens := []Token{tok(beginJavadoc, "/**")}
	for !l.input.isExhausted() {
		t, err := l.readToken()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, t)
	}
	if !l.contextStack.isEmpty() {
		return nil, &lexError{"unbalanced javadoc tags"}
	}
	tokens = append(tokens, tok(endJavadoc, "*/"))

	result := joinAdjacentLiteralsAndAdjacentWhitespace(tokens)
	result = inferParagraphTags(result)
	result = optionalizeSpacesAfterLinks(result)
	result = deindentPreCodeBlocks(result)
	return result, nil
}

func (l *javadocLexer) readToken() (Token, error) {
	kind, err := l.consumeToken()
	if err != nil {
		return Token{}, err
	}
	return tok(kind, l.input.readAndResetRecorded()), nil
}

func (l *javadocLexer) consumeToken() (Kind, error) {
	preserve := l.preserveExistingFormatting()

	if l.input.tryConsumeRegex(reClassicNewline) {
		l.somethingSinceNewln = false
		if preserve {
			return forcedNewline, nil
		}
		return whitespace, nil
	}
	if l.input.tryConsume(" ") || l.input.tryConsume("\t") {
		if preserve {
			return literal, nil
		}
		return whitespace, nil
	}

	if !l.somethingSinceNewln && l.input.tryConsumeRegex(reFooterTag) {
		if !l.contextStack.isEmpty() {
			return 0, &lexError{"unbalanced javadoc tags"}
		}
		l.somethingSinceNewln = true
		return footerJavadocTagStart, nil
	}
	l.somethingSinceNewln = true

	if l.input.tryConsumeRegex(reSnippetTagOpen) {
		if l.contextStack.containsAny(braceContexts) {
			l.contextStack.push(ctxBrace)
			return literal, nil
		}
		l.contextStack.push(ctxSnippet)
		return snippetBegin, nil
	}
	if l.input.tryConsumeRegex(reInlineTagOpen) {
		l.contextStack.push(ctxInlineTag)
		return literal, nil
	}
	if l.input.tryConsume("{") {
		if l.contextStack.containsAny(braceContexts) {
			l.contextStack.push(ctxBrace)
		}
		return literal, nil
	}
	if l.input.tryConsume("}") {
		if popped, ok := l.contextStack.popIfIn(braceContexts); ok && popped == ctxSnippet {
			return snippetEnd, nil
		}
		return literal, nil
	}

	if l.contextStack.containsAny(tagContexts) {
		l.mustConsume(reClassicLiteral)
		return literal, nil
	}

	if l.input.tryConsumeRegex(rePreOpen) {
		l.contextStack.push(ctxHTMLPre)
		return preserveOr(preserve, preOpen), nil
	}
	if l.input.tryConsumeRegex(rePreClose) {
		l.contextStack.popUntil(ctxHTMLPre)
		return preserveOr(l.preserveExistingFormatting(), preClose), nil
	}
	if l.input.tryConsumeRegex(reCodeOpen) {
		l.contextStack.push(ctxHTMLCode)
		return preserveOr(preserve, codeOpen), nil
	}
	if l.input.tryConsumeRegex(reCodeClose) {
		l.contextStack.popUntil(ctxHTMLCode)
		return preserveOr(l.preserveExistingFormatting(), codeClose), nil
	}
	if l.input.tryConsumeRegex(reTableOpen) {
		l.contextStack.push(ctxTable)
		return preserveOr(preserve, tableOpen), nil
	}
	if l.input.tryConsumeRegex(reTableClose) {
		l.contextStack.popUntil(ctxTable)
		return preserveOr(l.preserveExistingFormatting(), tableClose), nil
	}

	if preserve {
		l.mustConsume(reClassicLiteral)
		return literal, nil
	}

	switch {
	case l.input.tryConsumeRegex(reParagraphOpen):
		return paragraphOpen, nil
	case l.input.tryConsumeRegex(reParagraphClose):
		return paragraphClose, nil
	case l.input.tryConsumeRegex(reListOpen):
		return listOpen, nil
	case l.input.tryConsumeRegex(reListClose):
		return listClose, nil
	case l.input.tryConsumeRegex(reListItemOpen):
		return listItemOpen, nil
	case l.input.tryConsumeRegex(reListItemClose):
		return listItemClose, nil
	case l.input.tryConsumeRegex(reBlockquoteOpen):
		return blockquoteOpen, nil
	case l.input.tryConsumeRegex(reBlockquoteClose):
		return blockquoteClose, nil
	case l.input.tryConsumeRegex(reHeaderOpen):
		return headerOpen, nil
	case l.input.tryConsumeRegex(reHeaderClose):
		return headerClose, nil
	case l.input.tryConsumeRegex(reBr):
		return br, nil
	case l.input.tryConsumeRegex(reMoeBegin):
		return moeBeginStrip, nil
	case l.input.tryConsumeRegex(reMoeEnd):
		return moeEndStrip, nil
	case l.input.tryConsumeRegex(reHTMLComment):
		return htmlComment, nil
	case l.input.tryConsumeRegex(reClassicLiteral):
		return literal, nil
	}
	return 0, &lexError{"javadoc lexer: no token matched"}
}

func preserveOr(preserve bool, k Kind) Kind {
	if preserve {
		return literal
	}
	return k
}

func (l *javadocLexer) mustConsume(p *regexp.Regexp) {
	if !l.input.tryConsumeRegex(p) {
		panic("javadoc lexer: expected literal")
	}
}

func (l *javadocLexer) preserveExistingFormatting() bool {
	return l.contextStack.containsAny(preserveFormatting)
}

func hasMultipleNewlines(s string) bool { return strings.Count(s, "\n") > 1 }

func joinAdjacentLiteralsAndAdjacentWhitespace(input []Token) []Token {
	var output []Token
	var accumulated strings.Builder
	i := 0
	for i < len(input) {
		if input[i].kind == literal {
			accumulated.WriteString(input[i].value)
			i++
			continue
		}
		if accumulated.Len() == 0 {
			output = append(output, input[i])
			i++
			continue
		}
		var seenWhitespace strings.Builder
		for i < len(input) && input[i].kind == whitespace {
			seenWhitespace.WriteString(input[i].value)
			i++
		}
		if i < len(input) && input[i].kind == literal && strings.HasPrefix(input[i].value, "@") {
			accumulated.WriteString(" ")
			accumulated.WriteString(input[i].value)
			i++
			continue
		}
		output = append(output, tok(literal, accumulated.String()))
		accumulated.Reset()
		if seenWhitespace.Len() > 0 {
			output = append(output, tok(whitespace, seenWhitespace.String()))
		}
	}
	return output
}

func inferParagraphTags(input []Token) []Token {
	var output []Token
	i := 0
	for i < len(input) {
		if input[i].kind == literal {
			output = append(output, input[i])
			i++
			if i < len(input) && input[i].kind == whitespace && hasMultipleNewlines(input[i].value) {
				output = append(output, input[i])
				i++
				if i < len(input) && input[i].kind == literal {
					output = append(output, tok(paragraphOpen, "<p>"))
				}
			}
		} else {
			output = append(output, input[i])
			i++
		}
	}
	return output
}

var reHref = regexp.MustCompile(`^href=[^>]*>$`)

func optionalizeSpacesAfterLinks(input []Token) []Token {
	var output []Token
	i := 0
	for i < len(input) {
		if input[i].kind == literal && reHref.MatchString(input[i].value) {
			output = append(output, input[i])
			i++
			if i < len(input) && input[i].kind == whitespace {
				output = append(output, tok(optionalLineBreak, input[i].value))
				i++
			}
		} else {
			output = append(output, input[i])
			i++
		}
	}
	return output
}

var reCodeFence = regexp.MustCompile(`^[ \t]*[{]@code$`)

func deindentPreCodeBlocks(input []Token) []Token {
	var output []Token
	i := 0
	for i < len(input) {
		if input[i].kind != preOpen {
			output = append(output, input[i])
			i++
			continue
		}
		output = append(output, input[i])
		i++
		var initialNewlines []Token
		for i < len(input) && input[i].kind == forcedNewline {
			initialNewlines = append(initialNewlines, input[i])
			i++
		}
		if i >= len(input) || input[i].kind != literal || !reCodeFence.MatchString(input[i].value) {
			output = append(output, initialNewlines...)
			if i < len(input) {
				output = append(output, input[i])
				i++
			}
			continue
		}
		output, i = deindentPreCodeBlock(output, input, i)
	}
	return output
}

func deindentPreCodeBlock(output []Token, input []Token, i int) ([]Token, int) {
	output = append(output, tok(literal, strings.TrimSpace(input[i].value)))
	i++
	var saved []Token
	for i < len(input) && input[i].kind != preClose {
		saved = append(saved, input[i])
		i++
	}
	for len(saved) > 0 && saved[0].kind == forcedNewline {
		saved = saved[1:]
	}
	for len(saved) > 0 && saved[len(saved)-1].kind == forcedNewline {
		saved = saved[:len(saved)-1]
	}
	if len(saved) == 0 {
		return output, i
	}

	last := saved[len(saved)-1]
	trailingBrace := false
	if last.kind == literal && strings.HasSuffix(last.value, "}") {
		saved = saved[:len(saved)-1]
		if len(last.value) > 1 {
			saved = append(saved, tok(literal, last.value[:len(last.value)-1]))
			saved = append(saved, tok(forcedNewline, ""))
		}
		trailingBrace = true
	}

	trim := -1
	for _, t := range saved {
		if t.kind == literal {
			idx := strings.IndexFunc(t.value, func(r rune) bool { return r != ' ' })
			if idx != -1 && (trim == -1 || idx < trim) {
				trim = idx
			}
		}
	}

	output = append(output, tok(forcedNewline, "\n"))
	for _, t := range saved {
		if t.kind == literal {
			if trim > 0 && len(t.value) > trim {
				output = append(output, tok(literal, t.value[trim:]))
			} else {
				output = append(output, t)
			}
		} else {
			output = append(output, t)
		}
	}
	if trailingBrace {
		output = append(output, tok(literal, "}"))
	} else {
		output = append(output, tok(forcedNewline, "\n"))
	}
	return output, i
}

File: src/format/javadoc/lexer.ts
// Port of google-java-format core/.../java/javadoc/JavadocLexer.java.
//
// Lexes a classic `/** ... */` javadoc comment into a token stream for the
// writer. The Markdown `///` path (MarkdownPositions) is deferred (phase 5):
// markdownTokensAt() returns nothing and classicJavadoc is always true here, so
// the markdown-only branches are inert.

import { CharStream } from "./char-stream.ts";
import { NestingStack } from "./nesting-stack.ts";
import { factory, type Token, tok } from "./token.ts";

export class LexException extends Error {}

const enum Ctx {
  HtmlPre,
  HtmlCode,
  MarkdownCode,
  Table,
  Snippet,
  Brace,
  InlineTag,
}

// Sticky/anchored patterns (the `y` flag matches only at the cursor).
const NON_UNIX_LINE_ENDING = /\r\n?/g;
const CLASSIC_NEWLINE = /[ \t]*\n[ \t]*[*]?[ \t]?/y;
const FOOTER_TAG = /@(?:param\s+<\w+>|[a-z]\w*)/y;
const MOE_BEGIN = /<!--\s*MOE:begin_intracomment_strip\s*-->/y;
const MOE_END = /<!--\s*MOE:end_intracomment_strip\s*-->/y;
const HTML_COMMENT = /<!--[\s\S]*?-->/y;
const SNIPPET_TAG_OPEN = /[{]@snippet\b/y;
const INLINE_TAG_OPEN = /[{]@\w*/y;
const CLASSIC_LITERAL = /[\s\S][^ \t\n@<{}*]*/y;

const openTag = (name: string): RegExp => new RegExp(`<(?:${name})\\b[^>]*>`, "iy");
const closeTag = (name: string): RegExp => new RegExp(`</(?:${name})\\b[^>]*>`, "iy");

const PRE_OPEN = openTag("pre");
const PRE_CLOSE = closeTag("pre");
const CODE_OPEN = openTag("code");
const CODE_CLOSE = closeTag("code");
const TABLE_OPEN = openTag("table");
const TABLE_CLOSE = closeTag("table");
const LIST_OPEN = openTag("ul|ol|dl");
const LIST_CLOSE = closeTag("ul|ol|dl");
const LIST_ITEM_OPEN = openTag("li|dt|dd");
const LIST_ITEM_CLOSE = closeTag("li|dt|dd");
const HEADER_OPEN = openTag("h[1-6]");
const HEADER_CLOSE = closeTag("h[1-6]");
const PARAGRAPH_OPEN = openTag("p");
const PARAGRAPH_CLOSE = closeTag("p");
const BLOCKQUOTE_OPEN = openTag("blockquote");
const BLOCKQUOTE_CLOSE = closeTag("blockquote");
const BR = openTag("br");

const TAG_CONTEXTS = [Ctx.Snippet, Ctx.InlineTag];
const BRACE_CONTEXTS = [Ctx.Snippet, Ctx.InlineTag, Ctx.Brace];
const PRESERVE_FORMATTING = [Ctx.HtmlPre, Ctx.Table, Ctx.HtmlCode, Ctx.Snippet];

/** Lex a `/** ... *​/` comment (including the delimiters) into tokens. */
export function lex(input: string): Token[] {
  input = input.replace(NON_UNIX_LINE_ENDING, "\n");
  // stripJavadocBeginAndEnd
  if (!input.startsWith("/**") || !input.endsWith("*/") || input.length <= 4) {
    throw new LexException(`not a /** */ comment: ${input}`);
  }
  const body = input.slice(3, input.length - 2);
  return new JavadocLexer(new CharStream(body)).generateTokens();
}

class JavadocLexer {
  private readonly contextStack = new NestingStack<Ctx>();
  private somethingSinceNewline = false;

  constructor(private readonly input: CharStream) {}

  generateTokens(): Token[] {
    const tokens: Token[] = [tok("beginJavadoc", "/**")];
    while (!this.input.isExhausted()) {
      tokens.push(this.readToken());
    }
    this.checkMatchingTags();
    tokens.push(tok("endJavadoc", "*/"));

    let result = joinAdjacentLiteralsAndAdjacentWhitespace(tokens);
    result = inferParagraphTags(result);
    result = optionalizeSpacesAfterLinks(result);
    result = deindentPreCodeBlocks(result);
    return result;
  }

  private readToken(): Token {
    const make = this.consumeToken();
    return make(this.input.readAndResetRecorded());
  }

  private consumeToken(): (value: string) => Token {
    const preserve = this.preserveExistingFormatting();

    if (this.input.tryConsumeRegex(CLASSIC_NEWLINE)) {
      this.somethingSinceNewline = false;
      return factory(preserve ? "forcedNewline" : "whitespace");
    }
    if (this.input.tryConsume(" ") || this.input.tryConsume("\t")) {
      // Literal in a preserved context prevents breaking a <pre> line.
      return factory(preserve ? "literal" : "whitespace");
    }

    if (!this.somethingSinceNewline && this.input.tryConsumeRegex(FOOTER_TAG)) {
      this.checkMatchingTags();
      this.somethingSinceNewline = true;
      return factory("footerJavadocTagStart");
    }
    this.somethingSinceNewline = true;

    if (this.input.tryConsumeRegex(SNIPPET_TAG_OPEN)) {
      if (this.contextStack.containsAny(BRACE_CONTEXTS)) {
        this.contextStack.push(Ctx.Brace);
        return factory("literal");
      }
      this.contextStack.push(Ctx.Snippet);
      return factory("snippetBegin");
    }
    if (this.input.tryConsumeRegex(INLINE_TAG_OPEN)) {
      this.contextStack.push(Ctx.InlineTag);
      return factory("literal");
    }
    if (this.input.tryConsume("{")) {
      if (this.contextStack.containsAny(BRACE_CONTEXTS)) this.contextStack.push(Ctx.Brace);
      return factory("literal");
    }
    if (this.input.tryConsume("}")) {
      const popped = this.contextStack.popIfIn(BRACE_CONTEXTS);
      return factory(popped === Ctx.Snippet ? "snippetEnd" : "literal");
    }

    // Inside an inline tag, no HTML interpretation.
    if (this.contextStack.containsAny(TAG_CONTEXTS)) {
      this.mustConsume(CLASSIC_LITERAL);
      return factory("literal");
    }

    if (this.input.tryConsumeRegex(PRE_OPEN)) {
      this.contextStack.push(Ctx.HtmlPre);
      return factory(preserve ? "literal" : "preOpen");
    }
    if (this.input.tryConsumeRegex(PRE_CLOSE)) {
      this.contextStack.popUntil(Ctx.HtmlPre);
      return factory(this.preserveExistingFormatting() ? "literal" : "preClose");
    }
    if (this.input.tryConsumeRegex(CODE_OPEN)) {
      this.contextStack.push(Ctx.HtmlCode);
      return factory(preserve ? "literal" : "codeOpen");
    }
    if (this.input.tryConsumeRegex(CODE_CLOSE)) {
      this.contextStack.popUntil(Ctx.HtmlCode);
      return factory(this.preserveExistingFormatting() ? "literal" : "codeClose");
    }
    if (this.input.tryConsumeRegex(TABLE_OPEN)) {
      this.contextStack.push(Ctx.Table);
      return factory(preserve ? "literal" : "tableOpen");
    }
    if (this.input.tryConsumeRegex(TABLE_CLOSE)) {
      this.contextStack.popUntil(Ctx.Table);
      return factory(this.preserveExistingFormatting() ? "literal" : "tableClose");
    }

    if (preserve) {
      this.mustConsume(CLASSIC_LITERAL);
      return factory("literal");
    }

    if (this.input.tryConsumeRegex(PARAGRAPH_OPEN)) return factory("paragraphOpen");
    if (this.input.tryConsumeRegex(PARAGRAPH_CLOSE)) return factory("paragraphClose");
    if (this.input.tryConsumeRegex(LIST_OPEN)) return factory("listOpen");
    if (this.input.tryConsumeRegex(LIST_CLOSE)) return factory("listClose");
    if (this.input.tryConsumeRegex(LIST_ITEM_OPEN)) return factory("listItemOpen");
    if (this.input.tryConsumeRegex(LIST_ITEM_CLOSE)) return factory("listItemClose");
    if (this.input.tryConsumeRegex(BLOCKQUOTE_OPEN)) return factory("blockquoteOpen");
    if (this.input.tryConsumeRegex(BLOCKQUOTE_CLOSE)) return factory("blockquoteClose");
    if (this.input.tryConsumeRegex(HEADER_OPEN)) return factory("headerOpen");
    if (this.input.tryConsumeRegex(HEADER_CLOSE)) return factory("headerClose");
    if (this.input.tryConsumeRegex(BR)) return factory("br");
    if (this.input.tryConsumeRegex(MOE_BEGIN)) return factory("moeBeginStrip");
    if (this.input.tryConsumeRegex(MOE_END)) return factory("moeEndStrip");
    if (this.input.tryConsumeRegex(HTML_COMMENT)) return factory("htmlComment");
    if (this.input.tryConsumeRegex(CLASSIC_LITERAL)) return factory("literal");
    throw new Error("javadoc lexer: no token matched");
  }

  private mustConsume(p: RegExp): void {
    if (!this.input.tryConsumeRegex(p)) throw new Error("javadoc lexer: expected literal");
  }

  private preserveExistingFormatting(): boolean {
    return this.contextStack.containsAny(PRESERVE_FORMATTING);
  }

  private checkMatchingTags(): void {
    if (!this.contextStack.isEmpty()) throw new LexException("unbalanced javadoc tags");
  }
}

function hasMultipleNewlines(s: string): boolean {
  return (s.match(/\n/g)?.length ?? 0) > 1;
}

// Join adjacent literals (and adjacent whitespace), with the special case that a
// run of literals followed by whitespace + an `@`-literal is joined too (so a
// line break is not inserted before it, which would turn it into a tag).
function joinAdjacentLiteralsAndAdjacentWhitespace(input: Token[]): Token[] {
  const output: Token[] = [];
  let accumulated = "";
  let i = 0;
  const peek = (): Token | undefined => input[i];
  while (i < input.length) {
    if (peek()!.kind === "literal") {
      accumulated += input[i++].value;
      continue;
    }
    if (accumulated === "") {
      output.push(input[i++]);
      continue;
    }
    let seenWhitespace = "";
    while (peek()?.kind === "whitespace") seenWhitespace += input[i++].value;
    const p = peek();
    if (p && p.kind === "literal" && p.value.startsWith("@")) {
      accumulated += " ";
      accumulated += input[i++].value;
      continue;
    }
    output.push(tok("literal", accumulated));
    accumulated = "";
    if (seenWhitespace !== "") output.push(tok("whitespace", seenWhitespace));
    // leave the current token for the next iteration
  }
  return output;
}

// Insert a <p> between literals separated by a blank line. Must run after joining.
function inferParagraphTags(input: Token[]): Token[] {
  const output: Token[] = [];
  let i = 0;
  const peek = (): Token | undefined => input[i];
  while (i < input.length) {
    if (peek()!.kind === "literal") {
      output.push(input[i++]);
      if (peek()?.kind === "whitespace" && hasMultipleNewlines(peek()!.value)) {
        output.push(input[i++]);
        if (peek()?.kind === "literal") output.push(tok("paragraphOpen", "<p>"));
      }
    } else {
      output.push(input[i++]);
    }
  }
  return output;
}

// Replace whitespace after an `href=...>` literal with an optional line break.
function optionalizeSpacesAfterLinks(input: Token[]): Token[] {
  const output: Token[] = [];
  let i = 0;
  const peek = (): Token | undefined => input[i];
  while (i < input.length) {
    if (peek()!.kind === "literal" && /^href=[^>]*>$/.test(peek()!.value)) {
      output.push(input[i++]);
      if (peek()?.kind === "whitespace") output.push(tok("optionalLineBreak", input[i++].value));
    } else {
      output.push(input[i++]);
    }
  }
  return output;
}

// Adjust indentation inside `<pre>{@code` blocks: trim leading/trailing blank
// lines, de-indent to the least-indented line, move a trailing `}` to its own line.
function deindentPreCodeBlocks(input: Token[]): Token[] {
  const output: Token[] = [];
  let i = 0;
  const peek = (): Token | undefined => input[i];
  while (i < input.length) {
    if (peek()!.kind !== "preOpen") {
      output.push(input[i++]);
      continue;
    }
    output.push(input[i++]);
    const initialNewlines: Token[] = [];
    while (i < input.length && peek()!.kind === "forcedNewline") initialNewlines.push(input[i++]);
    const p = peek();
    if (!p || p.kind !== "literal" || !/^[ \t]*[{]@code$/.test(p.value)) {
      output.push(...initialNewlines);
      if (i < input.length) output.push(input[i++]);
      continue;
    }
    i = deindentPreCodeBlock(output, input, i);
  }
  return output;
}

function deindentPreCodeBlock(output: Token[], input: Token[], i: number): number {
  output.push(tok("literal", input[i++].value.trim()));
  const saved: Token[] = [];
  while (i < input.length && input[i].kind !== "preClose") saved.push(input[i++]);
  while (saved.length > 0 && saved[0].kind === "forcedNewline") saved.shift();
  while (saved.length > 0 && saved[saved.length - 1].kind === "forcedNewline") saved.pop();
  if (saved.length === 0) return i;

  // move a trailing `}` to its own line
  const last = saved[saved.length - 1];
  let trailingBrace = false;
  if (last.kind === "literal" && last.value.endsWith("}")) {
    saved.pop();
    if (last.value.length > 1) {
      saved.push(tok("literal", last.value.slice(0, -1)));
      saved.push(tok("forcedNewline", ""));
    }
    trailingBrace = true;
  }

  let trim = -1;
  for (const t of saved) {
    if (t.kind === "literal") {
      const idx = t.value.search(/[^ ]/);
      if (idx !== -1 && (trim === -1 || idx < trim)) trim = idx;
    }
  }

  output.push(tok("forcedNewline", "\n"));
  for (const t of saved) {
    if (t.kind === "literal") {
      output.push(
        tok("literal", trim > 0 && t.value.length > trim ? t.value.slice(trim) : t.value),
      );
    } else {
      output.push(t);
    }
  }
  output.push(trailingBrace ? tok("literal", "}") : tok("forcedNewline", "\n"));
  return i;
}
